TombEngine/Specific/Input/InputAction.cpp

```diff
@@ -0,0 +1,142 @@
+#include "framework.h"
+#include "Specific/Input/InputAction.h"
+
+#include "Renderer/Renderer11.h"
+#include "Specific/clock.h"
+
+using TEN::Renderer::g_Renderer;
+
+namespace TEN::Input
+{
+	InputAction::InputAction(ActionID actionID)
+	{
+		this->ID = actionID;
+	}
+
+	ActionID InputAction::GetID() const
+	{
+		return ID;
+	}
+
+	float InputAction::GetValue() const
+	{
+		return Value;
+	}
+
+	float InputAction::GetTimeActive() const
+	{
+		return TimeActive;
+	}
+
+	float InputAction::GetTimeInactive() const
+	{
+		return TimeInactive;
+	}
+
+	bool InputAction::IsClicked() const
+	{
+		return ((Value != 0.0f) && (PrevValue == 0.0f));
+	}
+
+	bool InputAction::IsHeld(float delayInSec) const
+	{
+		float delayInFrameTime = (delayInSec == 0.0f) ? 0.0f : round(delayInSec / DELTA_TIME);
+		return ((Value != 0.0f) && (TimeActive >= delayInFrameTime));
+	}
+
+	// To avoid desync on the second pulse, ensure initialDelayInSec is a multiple of delayInSec.
+	bool InputAction::IsPulsed(float delayInSec, float initialDelayInSec) const
+	{
+		if (this->IsClicked())
+			return true;
+
+		if (!this->IsHeld() || PrevTimeActive == 0.0f || TimeActive == PrevTimeActive)
+			return false;
+
+		float activeDelayInFrameTime = (TimeActive > round(initialDelayInSec / DELTA_TIME)) ? round(delayInSec / DELTA_TIME) : round(initialDelayInSec / DELTA_TIME);
+		float delayInFrameTime = std::floor(TimeActive / activeDelayInFrameTime) * activeDelayInFrameTime;
+		if (delayInFrameTime > (std::floor(PrevTimeActive / activeDelayInFrameTime) * activeDelayInFrameTime))
+			return true;
+
+		// Keeping version counting real time for future reference. -- Sezz 2022.10.01
```

Comment:
Can you elaborate on the meaning of this commented code? Why you have left it?

Reply:
The commented version relied on actual time passed (1 frame = 1 / 30), but because our delta time is a constant and can't correct for drift, occasionally a desync would occur and the pulse would stutter. To avoid this, `InputAction` counts frame time for the time being (1 frame = 1).

Reply:
I see